includes/class-wc-stripe-intent-controller.php

```diff
@@ -135,6 +136,81 @@ protected function handle_error( $e, $redirect_url ) {
 		wp_safe_redirect( $redirect_url );
 		exit;
 	}
+
+	/**
+	 * Creates a Setup Intent through AJAX while adding cards.
+	 */
+	public function create_setup_intent() {
+		if (
+			! is_user_logged_in()
+			|| ! isset( $_POST['stripe_source_id'] )
+			|| ! isset( $_POST['nonce'] )
+		) {
+			return;
+		}
+
+		try {
+			$source_id = wc_clean( $_POST['stripe_source_id'] );
+
+			// 1. Verify.
+			if (
+				! wp_verify_nonce( sanitize_key( $_POST['nonce'] ), 'wc_stripe_create_si' )
+				|| ! preg_match( '/^src_.*$/', $source_id )
+			) {
+				throw new Exception( __( 'Unable to verify your request. Please reload the page and try again.', 'woocommerce-gateway-stripe' ) );
+			}
+
+
+			// 2. Load the customer ID (and create a customer eventually).
+			$customer = new WC_Stripe_Customer( wp_get_current_user()->ID );
+
+			// 3. Attach the source to the customer (Setup Intents require that).
+			$source_object = $customer->attach_source( $source_id );
+			if ( is_wp_error( $source_object ) ) {
+				throw new Exception( $source_object->get_error_message() );
+			}
+
+			// 4. Generate the setup intent
+			$setup_intent = WC_Stripe_API::request(
+				[
+					'customer'       => $customer->get_id(),
+					'confirm'        => 'true',
+					'payment_method' => $source_id,
+				],
+				'setup_intents'
+			);
+
+			if ( $setup_intent->error ) {
+				$error_response_message = print_r( $setup_intent, true );
+				WC_Stripe_Logger::log("Failed create Setup Intent while saving a card.");
+				WC_Stripe_Logger::log("Response: $error_response_message");
+				throw new Exception( __( 'Your card could not be set up for future usage.', 'woocommerce-gateway-stripe' ) );
+			}
+
+			// 5. Respond.
+			if ( 'requires_action' === $setup_intent->status ) {
+				$response = [
+					'status'        => 'requires_action',
+					'client_secret' => $setup_intent->client_secret,
+				];
+			} else {
```

Comment:
Is there any possibility that `status` is something other than `requires_action` or `succeeded`? It's probably better to be explicit here, only return `'success'` if the Intent is successful, and log if it's in an unexpected state (or maybe throw an Exception?)

Reply:
The api lists the following possible statuses:

- `requires_payment_method`
- `requires_confirmation`
- `requires_action`
- `processing`
- `canceled`
- `succeeded`

`requires_payment_method` should not happen since that indicates a payment method is not attached (we're already supplying the source. The setup intent can return to this state after something fails (e.g. 3DS auth), but we haven't gone through that process at this point.

That means if we see `requires_payment_method` we need to return an error.

`requires_confirmation` can not happen, since we attempt confirm the setup intent when we create it.

We're already handling `requires_action`.

`processing` can only happen after additional actions (such as 3DS auth) are handled. There is the possibility of it taking a while to process the setup intent:

> While for certain payment methods (e.g., cards) processing can be quick, other payment methods can take up to a few days to process.
>
> [Source](https://stripe.com/docs/payments/intents#intent-statuses)

I guess that's something we might have to handle, but I'm honestly not sure how 😬 . I think for now we can think of this as a success. Maybe open an issue to look into this?

`succeeded` means we're successful.

Reply:
And nice catch btw @DanReyLop 🥇!

Reply:
Oh, and we'd need to set the `cancelled` status ourselves by canceling the setup intent. I guess that probably won't happen?

Reply:
I added some error handling in 4a55721a6693d6542ee66809c0d98ebd60752824, @DanReyLop do you think it's sufficient, or should we include more detailed error messages for each of the different statuses?

Reply:
> processing can only happen after additional actions (such as 3DS auth) are handled. There is the possibility of it taking a while to process the setup intent:

I guess we could treat it as `succeeded` for the purposes of this function?

> I added some error handling in 4a55721, @DanReyLop do you think it's sufficient, or should we include more detailed error messages for each of the different statuses?

Thanks for the rundown! It sounds like every status that's not `succeeded` or `requires_action` means that something has gone really wrong (except for `processing` which is a wildcard, that's the first time I've read about it). I see that's exactly what you added in 4a55721, so that amount of error handling is fine by me 👍